### PR TITLE
Feature/app page module

### PR DESCRIPTION
**Pull request by sunnylochun**

### CODE REVIEW

**Comment by sunnylochun** [comment]
[(5) 【頁面管理】後台資料串接 [3]](https://trello.com/c/OkEv0lRb/2157-5-%E3%80%90%E9%A0%81%E9%9D%A2%E7%AE%A1%E7%90%86%E3%80%91%E5%BE%8C%E5%8F%B0%E8%B3%87%E6%96%99%E4%B8%B2%E6%8E%A5-3)

File: src/hooks/appPage.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+export type AppPageProps = {
+  id: string
+  path: string | null
+  title: string | null
+  craftData: { [key: string]: string } | null
+  editorId: string | null
+  editorName: string | null
+  publishedAt: Date | null
+  updatedAt: Date | null
```
Comment (futianshen) [inline src/hooks/appPage.ts:14]: is not nullable in db

File: src/hooks/appPage.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+      variables: {
+        pageId: values.pageId,
+        updated: {
+          ...(values.path ? { path: values.path } : {}),
+          ...(values.title ? { title: values.title } : {}),
+          ...(values.editorId ? { editor_id: values.editorId } : {}),
+          ...(values.publishedAt !== undefined ? { published_at: values.publishedAt } : {}),
+          ...(values.craftData ? { craft_data: values.craftData } : {}),
+          is_deleted: values.isDeleted,
+        },
```
Comment (futianshen) [inline src/hooks/appPage.ts:180]: nice pattern to avoid mutating other columns!

File: src/containers/ElementConfig.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { AppProvider } from 'lodestar-app-element/src/contexts/AppContext'
+import { AuthProvider } from 'lodestar-app-element/src/contexts/AuthContext'
+import React from 'react'
+import { useAuth } from '../contexts/AuthContext'
+
+export const ElementConfig: React.FC<{
```
Comment (futianshen) [inline src/containers/ElementConfig.tsx:6]: I think this is also a context, not a container(HOC)
maybe `ElementProvider` is a better naming

File: src/hooks/appPage.ts (hunk 2 above)
Comment (kkshyu) [reply src/hooks/appPage.ts:180]:
```js
const removeUndefinedValue = obj => JSON.parse(JSON.stringify(obj))
```

File: src/pages/CraftPage/CraftSettingsPanel.tsx [hunk 4]
```diff
@@ -4,9 +4,15 @@ import React from 'react'
 import { useIntl } from 'react-intl'
 import { craftPageMessages } from '../../helpers/translation'
 
-const CraftSettingsPanel: React.VFC = () => {
+const CraftSettingsPanel: React.VFC<{ setActiveKey: React.Dispatch<React.SetStateAction<string>> }> = ({
+  setActiveKey,
```
Comment (futianshen) [inline src/pages/CraftPage/CraftSettingsPanel.tsx:8]: NITS: can just use a callback to set activity key to `component`

File: src/pages/CraftPage/CraftSettingsPanel.tsx [hunk 5]
```diff
@@ -4,9 +4,13 @@ import React from 'react'
 import { useIntl } from 'react-intl'
 import { craftPageMessages } from '../../helpers/translation'
 
-const CraftSettingsPanel: React.VFC = () => {
+const CraftSettingsPanel: React.VFC<{ callback?: { onDelete?: () => void } }> = ({ callback }) => {
```
Comment (futianshen) [inline src/pages/CraftPage/CraftSettingsPanel.tsx:7]: can directly use `onDelete` as  callback props

File: src/hooks/appPage.ts (hunk 2 above)
Comment (kkshyu) [reply src/hooks/appPage.ts]:
```graphql
mutation UPDATE_APP(_set: $updated) {
}
```

```js
{
   variables: {
      updated: {
          title: 'foo'
          description: undefined
      }
   }
}
```

```js
const { update } = useProgram(programId)
update({ title: "xxx" })
```

File: src/pages/CraftPage/CraftSettingsPanel.tsx (hunk 5 above)
Comment (kkshyu) [reply src/pages/CraftPage/CraftSettingsPanel.tsx]:
```js
deleteCallbacks: {
    onDeleted: () => {}
    onBeforeDelete: () => {}
}
```

**Pull request merged**